internal/export/worker.go

```diff
@@ -57,93 +58,115 @@ func (s *Server) handleDoWork() http.Handler {
 		ctx := r.Context()
 
 		logger := logging.FromContext(ctx).Named("handleDoWork")
+		logger.Debugw("starting do-work worker")
+		defer logger.Debugw("finished do-work worker")
 
-		ctx, cancel := context.WithTimeout(r.Context(), s.config.WorkerTimeout)
+		ctx, cancel := context.WithTimeout(ctx, s.config.WorkerTimeout)
 		defer cancel()
 
-		indexWrittenForConfig := make(map[int64]struct{})
+		var merr *multierror.Error
+
+		indexesWritten := make(map[int64]struct{})
+
+	OUTER:
 		for {
 			if ctx.Err() != nil {
-				msg := "Timed out processing batches. Will continue on next invocation."
-				logger.Info(msg)
-				fmt.Fprintln(w, msg)
-				return
+				logger.Warnw("deadline passed, still work to do")
+				merr = multierror.Append(merr, fmt.Errorf("deadline exceeded"))
+				break OUTER
 			}
 
 			// Check for a batch and obtain a lease for it.
 			batch, err := exportdatabase.New(db).LeaseBatch(ctx, s.config.WorkerTimeout, time.Now())
 			if err != nil {
 				logger.Errorw("failed to lease batch", "error", err)
-				continue
```

Comment:
I think this was a bug. This could cause the worker to sit in a hot loop for 600s (timeout) if the database was unavailable. I think returning on the first instance of this error is the correct behavior? Otherwise, we should at least sleep for a second to prevent resource exhaustion.